wp-includes/service-workers.php

```diff
@@ -0,0 +1,121 @@
+<?php
+/**
+ * Service Worker functions.
+ *
+ * @since ?
+ *
+ * @package PWA
+ */
+
+/**
+ * Initialize $wp_service_workers if it has not been set.
+ *
+ * @global WP_Service_Workers $wp_service_workers
+ * @return WP_Service_Workers WP_Service_Workers instance.
+ */
+function wp_service_workers() {
+	global $wp_service_workers;
+	if ( ! ( $wp_service_workers instanceof WP_Service_Workers ) ) {
+		$wp_service_workers = new WP_Service_Workers();
+	}
+	return $wp_service_workers;
+}
+
+/**
+ * Register a new service worker.
+ *
+ * @since ?
+ *
+ * @param string $handle Name of the service worker. Should be unique.
+ * @param string $path   Relative path of the service worker.
+ * @param array  $deps   Optional. An array of registered script handles this depends on. Default empty array.
+ * @param string $scope  Optional Scope of the service worker.
+ * @return bool Whether the script has been registered. True on success, false on failure.
+ */
+function wp_register_service_worker( $handle, $path, $deps = array(), $scope = null ) {
+	$wp_service_workers = wp_service_workers();
+
+	// If the path is not correct.
+	if ( is_wp_error( $wp_service_workers->get_validated_file_path( $path ) ) ) {
+		_doing_it_wrong(
+			__FUNCTION__,
+			/* translators: %s is file URL */
+			sprintf( esc_html__( 'Service worker path is incorrect: %s', 'pwa' ), esc_url( $path ) ),
+			'0.1'
+		);
+	}
+
+	$registered = $wp_service_workers->add( $handle, $path, $deps, false, $scope );
+
+	return $registered;
+}
+
+/**
+ * Get service worker URL by scope.
+ *
+ * @param string $scope Scope, for example 'wp-admin'.
+ * @return string Service Worker URL.
+ */
+function wp_get_service_worker_url( $scope ) {
+
+	if ( ! $scope ) {
+		$scope = site_url( '/', 'relative' );
+	}
+
+	if ( get_option( 'permalink_structure' ) ) {
+		return add_query_arg( compact( 'scope' ), site_url( '/wp-service-worker.js', 'relative' ) );
```

Comment:
Do we actually need a separate case and rewrite for pretty permalinks at all? This URL wouldn't actually be directly used anyway in browser. Thoughts?

Reply:
You're right. If we can avoid the rewrite rules then we can save headaches. I think it is fine to just use the non-pretty URLs.